### PR TITLE
feat: date/timestamp bound check

### LINKED ISSUE #4874: Datetime type `attempt to multiply with overflow` from chrono
**Summary**

```
create table t1(a datetime);
insert into t1 values('9999-12-31 23:59:59.999999');

ERROR 2013 (HY000): Lost connection to MySQL server during query
```

Server error log:
```
2022-04-15T08:14:41.711788Z ERROR common_tracing::panic_hook: panicked at 'attempt to multiply with overflow', /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/chrono-0.4.19/src/naive/datetime.rs:351:21 backtrace=   0: common_tracing::panic_hook::set_panic_hook::{{closure}}
             at common/tracing/src/panic_hook.rs:25:25
   1: std::panicking::rust_panic_with_hook
             at /rustc/76d770ac21d9521db6a92a48c7b3d5b2cc535941/library/std/src/panicking.rs:702:17
   2: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/76d770ac21d9521db6a92a48c7b3d5b2cc535941/library/std/src/panicking.rs:586:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/76d770ac21d9521db6a92a48c7b3d5b2cc535941/library/std/src/sys_common/backtrace.rs:138:18
   4: rust_begin_unwind
             at /rustc/76d770ac21d9521db6a92a48c7b3d5b2cc535941/library/std/src/panicking.rs:584:5
   5: core::panicking::panic_fmt
             at /rustc/76d770ac21d9521db6a92a48c7b3d5b2cc535941/library/core/src/panicking.rs:143:14
   6: core::panicking::panic
             at /rustc/76d770ac21d9521db6a92a48c7b3d5b2cc535941/library/core/src/panicking.rs:48:5
   7: chrono::naive::datetime::NaiveDateTime::timestamp_nanos
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/chrono-0.4.19/src/naive/datetime.rs:351:21
   8: chrono::datetime::DateTime<Tz>::timestamp_nanos
             at /home/bohu/.cargo/registry/src/github.com-1ecc6299db9ec823/chrono-0.4.19/src/datetime.rs:164:9
   9: <common_datavalues::types::deserializations::date_time::DateTimeDeserializer<T> as common_datavalues::types::deserializations::TypeDeserializer>::de_text_quoted
             at common/datavalues/src/types/deserializations/date_time.rs:77:35
  10: databend_query::sql::statements::value_source::ValueSource::stream_read
             at query/src/sql/statements/value_source.rs:81:17
  11: databend_query::sql::statements::statement_insert::DfInsertStatement::analyze_insert_values::{{closure}}
             at query/src/sql/statements/statement_insert.rs:150:27
  12: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/76d770ac21d9521db6a92a48c7b3d5b2cc535941/library/core/src/future/mod.rs:91:19
  13: <databend_query::sql::statements::statement_insert::DfInsertStatement as databend_query::sql::statements::analyzer_statement::AnalyzableStatement>::analyze::{{closure}}::{{closure}}
             at query/src/sql/statements/statement_insert.rs:88:78
  14: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/76d770ac21d9521db6a92a48c7b3d5b2cc535941/library/core/src/future/mod.rs:91:19
  15: <databend_query::sql::statements::statement_insert::DfInsertStatement as databend_query::sql::statements::analyzer_statement::AnalyzableStatement>::analyze::{{closure}}
             at query/src/sql/statements/statement_insert.rs:76:5
  16: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/76d770ac21d9521db6a92a48c7b3d5b2cc535941/library/core/src/future/mod.rs:91:19
  17: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/76d770ac21d9521db6a92a48c7b3d5b2cc535941/library/core/src/future/future.rs:124:9
  18: <databend_query::sql::sql_statement::DfStatement as databend_query::sql::statements::analyzer_statement::AnalyzableStatement>::analyze::{{closure}}
             at query/src/sql/statements/analyzer_statement.rs:171:58

```

### PR DESCRIPTION
**Pull request by Veeupup**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

date/timestamp bound check
* precison only support 0~6
* date ranges from `1000-01-01` to `9999-12-31`
* timestamp ranges from `1000-01-01 00:00:00` to `9999-12-31 23:59:59`

## Changelog

- Feature

## Related Issues

Fixes #4874

### CODE REVIEW

**Comment by BohuTANG** [comment]
Hmm, we also need bound checks for datetime functions, like `select subtractYears(totimestamp(18875), 2);` will panic the server.

**Comment by Veeupup** [comment]
@BohuTANG  yes, it has been done and some stateless-test has been added.

File: common/datavalues/src/types/type_timestamp.rs [hunk 1]
```diff
@@ -56,21 +73,21 @@ impl TimestampType {
 
     #[inline]
     pub fn utc_timestamp(&self, v: i64) -> DateTime<Utc> {
-        let v = v * 10_i64.pow(9 - self.precision as u32);
+        let v = v * 10_i64.pow(6 - self.precision as u32);
```
Comment (BohuTANG) [inline common/datavalues/src/types/type_timestamp.rs:76]: Where did the check `self.precision <= 6`?

File: common/datavalues/src/types/deserializations/date.rs [hunk 2]
```diff
@@ -35,6 +35,7 @@ where
 {
     fn de_binary(&mut self, reader: &mut &[u8]) -> Result<()> {
         let value: T = reader.read_scalar()?;
```
Comment (sundy-li) [inline common/datavalues/src/types/deserializations/date.rs:37]: `T` is no longer needed, it can be remove in another pr.

File: common/datavalues/src/types/type_timestamp.rs (hunk 1 above)
Comment (BohuTANG) [reply common/datavalues/src/types/type_timestamp.rs:76]: I think we should check it in datatype, it's datatype limit not sql.

File: common/datavalues/src/types/type_timestamp.rs [hunk 3]
```diff
@@ -106,9 +123,6 @@ impl DataType for TimestampType {
             4 => &["DateTime(4)"],
             5 => &["DateTime(5)"],
             6 => &["Timestamp", "DateTime"],
-            7 => &["DateTime(7)"],
-            8 => &["DateTime(8)"],
-            9 => &["DateTime(9)"],
```
Comment (b41sh) [inline common/datavalues/src/types/type_timestamp.rs:111]: Why we don't support precision 9?

Comment (Veeupup) [reply common/datavalues/src/types/type_timestamp.rs:111]: for two reasons, one is cockroachDB and duckdb only support timestamp with precision microsecond.
another is that is our timestamp bound of ts is `9999-12-31 23:59:59` whose nanosecond is beyond i64::MAX and to make it simple, we can store the total timestamp in a `i64`

Comment (b41sh) [reply common/datavalues/src/types/type_timestamp.rs:111]: ok

File: common/datavalues/src/types/type_timestamp.rs (hunk 1 above)
Comment (Veeupup) [reply common/datavalues/src/types/type_timestamp.rs:76]: yes, we only register timestamp datatype with precision from 0 to 6
https://github.com/Veeupup/databend/blob/a4c2671481/common/datavalues/src/types/type_factory.rs#L53-L63
